peps/pep-0788.rst

```diff
@@ -0,0 +1,817 @@
+        if (PyThreadState_Ensure(interp) < 0) {
+            fputs("Cannot talk to Python", stderr);
+            return -1;
+        }
+        (void)PyThreadState_SetDaemon(1);
+        if (PyRun_SimpleString("print(42)") < 0) {
+            PyErr_Print();
+        }
+        PyThreadState_Release();
+        return 0;
+    }
+
+    static PyObject *
+    my_method(PyObject *self, PyObject *unused)
+    {
+        PyThread_handle_t handle;
+        PyThead_indent_t indent;
+
+        PyInterpreterState *interp = PyInterpreterState_Hold();
+        if (PyThread_start_joinable_thread(thread_func, interp, &ident, &handle) < 0) {
+            PyInterpreterState_Release(interp);
+            return NULL;
+        }
+        Py_RETURN_NONE;
+    }
+
+Asynchronous callback example
+*****************************
+
+As started in the Motivation_, there are many cases where it's desirable
+to call Python in an asynchronous callback, such as a signal handler. In that
+case, it's not safe to call :c:func:`PyInterpreterState_Hold`, because it's
+not guaranteed that :c:func:`PyThreadState_Ensure` will ever be called, which
+would deadlock finalization.
+
+This scenario requires :c:func:`PyInterpreterState_Lookup`, which only prevents
+finalization when the lookup has been made.
+
+For example:
+
+.. code-block:: c
+
+    typedef struct {
+        int64_t interp_id;
+    } pyrun_t;
+
+    static int
+    async_callback(void *arg)
+    {
+        pyrun_t *data = (pyrun_t *)arg;
+        PyInterpreterState *interp = PyInterpreterState_Lookup(data->interp_id);
+        PyMem_RawFree(data);
+        if (interp == NULL) {
+            fputs("Python has shut down", stderr);
+            return -1;
+        }
+        if (PyThreadState_Ensure(interp) < 0) {
+            fputs("Cannot talk to Python", stderr);
+            return -1;
+        }
+        if (PyRun_SimpleString("print(42)") < 0) {
+            PyErr_Print();
+        }
+        PyThreadState_Release();
+        return 0;
+    }
+
+    static PyObject *
+    setup_callback(PyObject *self, PyObject *unused)
+    {
+        PyThread_handle_t handle;
+        PyThead_indent_t indent;
+
+        pyrun_t *data = PyMem_RawMalloc(sizeof(pyrun_t));
+        if (data == NULL) {
+            return PyErr_NoMemory();
+        }
+        // Weak reference to the interpreter. It won't wait on the callback
+        // to finalize.
+        data->interp_id = PyInterpreterState_GetID(PyInterpreterState_Get());
+        register_callback(async_callback, data);
+
+        Py_RETURN_NONE;
+    }
+
+Reference Implementation
+========================
+
+A reference implementation of this PEP can be found
+`here <https://github.com/ZeroIntensity/cpython/tree/pep-788-impl>`_.
+
+Rejected Ideas
+==============
+
+Using an interpreter ID instead of a interpreter state for :c:func:`PyThreadState_Ensure`
+-----------------------------------------------------------------------------------------
```

Comment:
```suggestion
Using an interpreter ID instead of a interpreter state for ``PyThreadState_Ensure``
-----------------------------------------------------------------------------------
```